src/libpriv/rpmostree-rpm-util.cxx

```diff
@@ -1043,80 +1043,36 @@ rpmostree_print_transaction (DnfContext *dnfctx)
     rpmostree_output_message ("Empty transaction");
 }
 
-struct _cap_struct
-{
-  struct __user_cap_header_struct head;
-  union
-  {
-    struct __user_cap_data_struct set;
-    __u32 flat[3];
-  } u[_LINUX_CAPABILITY_U32S_2];
-};
-
-/* Rewritten version of _fcaps_save from libcap, since it's not
- * exposed, and we need to generate the raw value.
- */
-static void
-cap_t_to_vfs (cap_t cap_d, struct vfs_cap_data *rawvfscap, int *out_size)
-{
-  guint32 eff_not_zero, magic;
-  guint tocopy, i;
-
-  /* Hardcoded to 2.  There is apparently a version 3 but it just maps
-   * to 2.  I doubt another version would ever be implemented, and
-   * even if it was we'd need to be backcompatible forever.  Anyways,
-   * setuid/fcaps binaries should go away entirely.
-   */
-  magic = VFS_CAP_REVISION_2;
-  tocopy = VFS_CAP_U32_2;
-  *out_size = XATTR_CAPS_SZ_2;
-
-  for (eff_not_zero = 0, i = 0; i < tocopy; i++)
-    eff_not_zero |= cap_d->u[i].flat[CAP_EFFECTIVE];
-
-  /* Here we're also not validating that the kernel understands
-   * the capabilities.
-   */
-
-  for (i = 0; i < tocopy; i++)
-    {
-      rawvfscap->data[i].permitted = GUINT32_TO_LE (cap_d->u[i].flat[CAP_PERMITTED]);
-      rawvfscap->data[i].inheritable = GUINT32_TO_LE (cap_d->u[i].flat[CAP_INHERITABLE]);
-    }
-
-  if (eff_not_zero == 0)
-    rawvfscap->magic_etc = GUINT32_TO_LE (magic);
-  else
-    rawvfscap->magic_etc = GUINT32_TO_LE (magic | VFS_CAP_FLAGS_EFFECTIVE);
-}
-
-// Given a file capability string, convert it to the convention ostree uses
-// to an store extended attribute value: (@ay@ay)
 GVariant *
-rpmostree_fcap_to_ostree_xattr (const char *fcap)
+rpmostree_fcap_to_ostree_xattr (const char *fcap, GError **error)
 {
+  /* Unfortunately, libcap doesn't expose any APIs to get the raw xattr value.
+   * For now, we hackily dance around this by writting it out to a file and then
+   * re-reading from it. */
   cap_t caps = cap_from_text (fcap);
-  struct vfs_cap_data vfscap = {
-    0,
-  };
 
-  int vfscap_size;
-  cap_t_to_vfs (caps, &vfscap, &vfscap_size);
-  cap_free (caps);
-  g_autoptr (GBytes) vfsbytes = g_bytes_new (&vfscap, vfscap_size);
+  g_auto (GLnxTmpfile) tmpf = { 0 };
+  if (!glnx_open_anonymous_tmpfile (O_WRONLY, &tmpf, error))
```

Comment:
Missing `O_CLOEXEC`

Reply:
Added! (Though I don't think it matters much in this context, no?)

Reply:
When we spawn processes with glib, it closes all file descriptors by default.  Rust's `std` doesn't!  Which means we're subject to potentially leaking these things unintentionally.

We basically *always* want `O_CLOEXEC` and only to unset the cloexec flag in fork helpers.

Reply:
Interesting, hadn't realized that re. glib vs Rust.

I was more referring to the fact that this is a short-lived fd anyway and that when this is called, I don't think we spawn things on any thread. But certainly, `O_CLOEXEC` is warranted even just as good practice.